src/libstore/local-overlay-store.cc

```diff
@@ -182,7 +182,7 @@ void LocalOverlayStore::addToStore(const ValidPathInfo & info, Source & source,
     LocalStore::addToStore(info, source, repair, checkSigs);
     if (lowerStore->isValidPath(info.path)) {
         // dedup stores
-        deletePath(toUpperPath(info.path));
+        deletePath(toUpperPath(info.path));  // TODO: Investigate whether this can trigger 'stale file handle' errors.
```

Comment:
Good catch. I think we can delete this actually, because `LocalOverlayStore::isValidPath` more recently will put in the upper DB, and `LocalStore::addToStore` will call that.

Reply:
If so, we can stop overriding this function altogether.

Reply:
Do we need to change `isValidPathUncached` to do the deduplication then?

Agreed on your main point though. It's cleaner to maintain the necessary invariants solely in the query methods, and avoid overriding the insert/update methods altogether.

Reply:
I removed the `add` methods for now. We can figure out the deduplication later when we're getting GC working properly.